File: backend/app/schemas/crowd.py
"""Crowd 검수 API schemas — Phase 2.2.10 placeholder + Phase 4.2.1 정식.

Phase 4 의 정식 schema 는 `crowd.task` 기반. Phase 2.2.10 의 `CrowdTaskOut` 등은
`run.crowd_task` view 호환 형태로 유지 (legacy endpoint 호환).
"""

from __future__ import annotations

from datetime import date, datetime
from typing import Any, Literal

from pydantic import BaseModel, ConfigDict, Field

# Phase 2.2.10 호환 (run.crowd_task view 가 같은 형태) — 기존 endpoint 가 사용.
CrowdTaskStatus = Literal["PENDING", "REVIEWING", "APPROVED", "REJECTED"]


class CrowdTaskOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    crowd_task_id: int
    raw_object_id: int
    partition_date: date
    ocr_result_id: int | None
    reason: str
    status: CrowdTaskStatus
    payload_json: dict[str, Any]
    assigned_to: int | None
    created_at: datetime
    reviewed_at: datetime | None
    reviewed_by: int | None


class OcrResultPreview(BaseModel):
    ocr_result_id: int
    page_no: int | None
    text_content: str | None
    confidence_score: float | None
    engine_name: str


class CrowdTaskDetail(CrowdTaskOut):
    """단건 조회 — raw_object payload + ocr_result 본문 포함."""

    raw_object_uri: str | None = None
    raw_object_payload: dict[str, Any] | None = None
    ocr_results: list[OcrResultPreview] = Field(default_factory=list)


class CrowdTaskStatusUpdate(BaseModel):
    status: Literal["REVIEWING", "APPROVED", "REJECTED"]
    note: str | None = Field(default=None, max_length=500)


# ---------------------------------------------------------------------------
# Phase 4.2.1 — 정식 crowd schema 기반 DTO
# ---------------------------------------------------------------------------
TaskKind = Literal[
    "OCR_REVIEW",
    "PRODUCT_MATCHING",
    "RECEIPT_VALIDATION",
    "ANOMALY_CHECK",
    "std_low_confidence",
    "ocr_low_confidence",
    "price_fact_low_confidence",
    "sample_review",
]
TaskStatus = Literal["PENDING", "REVIEWING", "CONFLICT", "APPROVED", "REJECTED", "CANCELLED"]
ReviewDecision = Literal["APPROVE", "REJECT", "SKIP"]
ConsensusKind = Literal["SINGLE", "DOUBLE_AGREED", "CONFLICT_RESOLVED"]


class TaskOut(BaseModel):
    """crowd.task 1행."""

    model_config = ConfigDict(from_attributes=True)

    crowd_task_id: int
    task_kind: TaskKind
    priority: int
    raw_object_id: int | None
    partition_date: date | None
    ocr_result_id: int | None
    std_record_id: int | None
    payload: dict[str, Any]
    status: TaskStatus
    requires_double_review: bool
    created_at: datetime
    updated_at: datetime


class TaskAssignmentOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    assignment_id: int
    crowd_task_id: int
    reviewer_id: int
    assigned_at: datetime
    due_at: datetime | None
    released_at: datetime | None


class ReviewOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    review_id: int
    crowd_task_id: int
    reviewer_id: int
    decision: ReviewDecision
    decision_payload: dict[str, Any]
    comment: str | None
    time_spent_ms: int | None
    decided_at: datetime


class TaskDecisionOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    crowd_task_id: int
    final_decision: Literal["APPROVE", "REJECT"]
    decided_by: int | None
    consensus_kind: ConsensusKind
    effect_payload: dict[str, Any]
    decided_at: datetime


class TaskFullDetail(TaskOut):
    """단건 상세 — assignments + reviews + decision 동봉."""

    assignments: list[TaskAssignmentOut] = Field(default_factory=list)
    reviews: list[ReviewOut] = Field(default_factory=list)
    decision: TaskDecisionOut | None = None


class AssignTaskRequest(BaseModel):
    """검수자 1+명 배정. priority>=8 면 자동으로 2명 이상 강제."""

    reviewer_ids: list[int] = Field(min_length=1, max_length=5)
    due_at: datetime | None = None


class SubmitReviewRequest(BaseModel):
    """검수자가 자신의 결정을 제출."""

    decision: ReviewDecision
    decision_payload: dict[str, Any] = Field(default_factory=dict)
    comment: str | None = Field(default=None, max_length=2000)
    time_spent_ms: int | None = Field(default=None, ge=0, le=3_600_000)


class ResolveConflictRequest(BaseModel):
    """관리자 (ADMIN/APPROVER) 가 CONFLICT 해결."""

    final_decision: Literal["APPROVE", "REJECT"]
    note: str | None = Field(default=None, max_length=2000)


class ReviewerStatsOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    reviewer_id: int
    reviewed_count_30d: int
    avg_decision_ms_30d: int | None
    conflict_rate_30d: float | None
    regression_rate_30d: float | None
    updated_at: datetime


__all__ = [
    "AssignTaskRequest",
    "ConsensusKind",
    "CrowdTaskDetail",
    "CrowdTaskOut",
    "CrowdTaskStatus",
    "CrowdTaskStatusUpdate",
    "OcrResultPreview",
    "ResolveConflictRequest",
    "ReviewDecision",
    "ReviewOut",
    "ReviewerStatsOut",
    "SubmitReviewRequest",
    "TaskAssignmentOut",
    "TaskDecisionOut",
    "TaskFullDetail",
    "TaskKind",
    "TaskOut",
    "TaskStatus",
]

File: frontend/src/api/crowd.ts
import { useMutation, useQuery, useQueryClient } from "@tanstack/react-query";
import { apiRequest } from "./client";

export type CrowdTaskStatus = "PENDING" | "REVIEWING" | "APPROVED" | "REJECTED";

export const CROWD_REASONS = [
  "ocr_low_confidence",
  "std_low_confidence",
  "price_fact_low_confidence",
  "price_fact_sample_review",
] as const;
export type CrowdReason = (typeof CROWD_REASONS)[number];

export interface CrowdTask {
  crowd_task_id: number;
  raw_object_id: number;
  partition_date: string;
  ocr_result_id: number | null;
  reason: string;
  status: CrowdTaskStatus;
  payload_json: Record<string, unknown>;
  assigned_to: number | null;
  created_at: string;
  reviewed_at: string | null;
  reviewed_by: number | null;
}

export interface OcrResultPreview {
  ocr_result_id: number;
  page_no: number | null;
  text_content: string | null;
  confidence_score: number | null;
  engine_name: string;
}

export interface CrowdTaskDetail extends CrowdTask {
  raw_object_uri: string | null;
  raw_object_payload: Record<string, unknown> | null;
  ocr_results: OcrResultPreview[];
}

export interface ListCrowdTasksParams {
  status?: CrowdTaskStatus;
  reason?: string;
  limit?: number;
  offset?: number;
}

export function useCrowdTasks(params: ListCrowdTasksParams = {}) {
  return useQuery({
    queryKey: ["crowd-tasks", params],
    queryFn: () =>
      apiRequest<CrowdTask[]>("/v1/crowd-tasks", { params: { ...params } }),
  });
}

export function useCrowdTaskDetail(crowdTaskId: number | null) {
  return useQuery({
    queryKey: ["crowd-tasks", crowdTaskId],
    enabled: crowdTaskId != null,
    queryFn: () =>
      apiRequest<CrowdTaskDetail>(`/v1/crowd-tasks/${crowdTaskId}`),
  });
}

export function useUpdateCrowdTaskStatus() {
  const qc = useQueryClient();
  return useMutation({
    mutationFn: ({
      crowdTaskId,
      status,
      note,
    }: {
      crowdTaskId: number;
      status: "REVIEWING" | "APPROVED" | "REJECTED";
      note?: string;
    }) =>
      apiRequest<CrowdTask>(`/v1/crowd-tasks/${crowdTaskId}/status`, {
        method: "PATCH",
        body: { status, note },
      }),
    onSuccess: () => {
      qc.invalidateQueries({ queryKey: ["crowd-tasks"] });
      qc.invalidateQueries({ queryKey: ["crowd-v4"] });
    },
  });
}

// ---------------------------------------------------------------------------
// Phase 4.2.1 정식 API (/v1/crowd/tasks)
// ---------------------------------------------------------------------------
export type TaskKind =
  | "OCR_REVIEW"
  | "PRODUCT_MATCHING"
  | "RECEIPT_VALIDATION"
  | "ANOMALY_CHECK"
  | "std_low_confidence"
  | "ocr_low_confidence"
  | "price_fact_low_confidence"
  | "sample_review";

export type TaskStatus =
  | "PENDING"
  | "REVIEWING"
  | "CONFLICT"
  | "APPROVED"
  | "REJECTED"
  | "CANCELLED";

export type ReviewDecision = "APPROVE" | "REJECT" | "SKIP";
export type ConsensusKind = "SINGLE" | "DOUBLE_AGREED" | "CONFLICT_RESOLVED";

export interface TaskV4 {
  crowd_task_id: number;
  task_kind: TaskKind;
  priority: number;
  raw_object_id: number | null;
  partition_date: string | null;
  ocr_result_id: number | null;
  std_record_id: number | null;
  payload: Record<string, unknown>;
  status: TaskStatus;
  requires_double_review: boolean;
  created_at: string;
  updated_at: string;
}

export interface TaskAssignmentV4 {
  assignment_id: number;
  crowd_task_id: number;
  reviewer_id: number;
  assigned_at: string;
  due_at: string | null;
  released_at: string | null;
}

export interface ReviewV4 {
  review_id: number;
  crowd_task_id: number;
  reviewer_id: number;
  decision: ReviewDecision;
  decision_payload: Record<string, unknown>;
  comment: string | null;
  time_spent_ms: number | null;
  decided_at: string;
}

export interface TaskDecisionV4 {
  crowd_task_id: number;
  final_decision: "APPROVE" | "REJECT";
  decided_by: number | null;
  consensus_kind: ConsensusKind;
  effect_payload: Record<string, unknown>;
  decided_at: string;
}

export interface TaskFullDetail extends TaskV4 {
  assignments: TaskAssignmentV4[];
  reviews: ReviewV4[];
  decision: TaskDecisionV4 | null;
}

export interface ReviewerStat {
  reviewer_id: number;
  display_name: string;
  count_30d: number;
  avg_decision_ms: number | null;
}

export function useCrowdTasksV4(params: {
  status?: TaskStatus;
  task_kind?: TaskKind;
  limit?: number;
  offset?: number;
}) {
  return useQuery({
    queryKey: ["crowd-v4", "tasks", params],
    queryFn: () =>
      apiRequest<TaskV4[]>("/v1/crowd/tasks", { params: { ...params } }),
  });
}

export function useCrowdTaskDetailV4(taskId: number | null) {
  return useQuery({
    queryKey: ["crowd-v4", "tasks", taskId],
    enabled: taskId != null,
    queryFn: () =>
      apiRequest<TaskFullDetail>(`/v1/crowd/tasks/${taskId}`),
  });
}

export function useAssignReviewers() {
  const qc = useQueryClient();
  return useMutation({
    mutationFn: ({
      taskId,
      reviewer_ids,
      due_at,
    }: {
      taskId: number;
      reviewer_ids: number[];
      due_at?: string | null;
    }) =>
      apiRequest<TaskAssignmentV4[]>(
        `/v1/crowd/tasks/${taskId}/assign`,
        { method: "POST", body: { reviewer_ids, due_at: due_at ?? null } },
      ),
    onSuccess: () => qc.invalidateQueries({ queryKey: ["crowd-v4"] }),
  });
}

export function useSubmitReview() {
  const qc = useQueryClient();
  return useMutation({
    mutationFn: ({
      taskId,
      decision,
      decision_payload,
      comment,
      time_spent_ms,
    }: {
      taskId: number;
      decision: ReviewDecision;
      decision_payload?: Record<string, unknown>;
      comment?: string | null;
      time_spent_ms?: number | null;
    }) =>
      apiRequest<ReviewV4>(`/v1/crowd/tasks/${taskId}/review`, {
        method: "POST",
        body: {
          decision,
          decision_payload: decision_payload ?? {},
          comment: comment ?? null,
          time_spent_ms: time_spent_ms ?? null,
        },
      }),
    onSuccess: () => qc.invalidateQueries({ queryKey: ["crowd-v4"] }),
  });
}

export function useResolveConflict() {
  const qc = useQueryClient();
  return useMutation({
    mutationFn: ({
      taskId,
      final_decision,
      note,
    }: {
      taskId: number;
      final_decision: "APPROVE" | "REJECT";
      note?: string | null;
    }) =>
      apiRequest<TaskDecisionV4>(`/v1/crowd/tasks/${taskId}/resolve`, {
        method: "POST",
        body: { final_decision, note: note ?? null },
      }),
    onSuccess: () => qc.invalidateQueries({ queryKey: ["crowd-v4"] }),
  });
}

export function useReviewerStats() {
  return useQuery({
    queryKey: ["crowd-v4", "stats", "reviewers"],
    queryFn: () => apiRequest<ReviewerStat[]>("/v1/crowd/stats/reviewers"),
    staleTime: 30_000,
  });
}
